File: emm/testing_service/src/api/v1/endpoints/test_logs.py
"""Чтение/экспорт логов прогонов (§2.6, §8.4 плана миграции).

Открыто любому аутентифицированному актору (`AuthenticatedIdentity`) — то же
соображение, что и у остальных read-путей сервиса (каталог глобальных
переменных, список стендов): чтение лога не секрет и не завязано на
department-scope роль. Если позже понадобится более строгий гейт — заводить
его отдельным решением, не задним числом здесь.
"""

from fastapi import APIRouter, Depends, Query, Response
from sqlalchemy.ext.asyncio import AsyncSession

from src.dependencies.auth import AuthenticatedIdentity
from src.dependencies.db import get_db
from src.schemas.common import PaginatedResponse
from src.schemas.test_log import TestLogSegmentResponse
from src.services import test_log as svc

router = APIRouter(prefix="/queue-items")


@router.get(
    "/{queue_item_id}/log",
    summary="Скачать/просмотреть текст лога прогона",
    description=(
        "Без `from`/`to` — весь текст лога, отдаётся как attachment с "
        "человекочитаемым именем файла (§8.3: "
        "`<stand>_<testname>_<os_version>_<kernel>_<date>.log`). С одним или "
        "обоими параметрами — диапазон текста по офсетам "
        "(`test_log_segments.byte_offset_*`) — для клика по сегменту в UI. "
        "Невалидный диапазон (`from>to`, за пределами длины текста) — 422."
    ),
    responses={
        401: {"description": "ACCESS_TOKEN_MISSING — запрос без bearer'а."},
        404: {"description": "TEST_LOG_NOT_FOUND — для этого queue_item лога ещё нет."},
        422: {"description": "LOG_RANGE_INVALID — невалидный from/to."},
    },
)
async def get_log(
    queue_item_id: str,
    identity: AuthenticatedIdentity,
    db: AsyncSession = Depends(get_db),
    from_: int | None = Query(default=None, ge=0, alias="from", description="Начало диапазона (включительно)."),
    to_: int | None = Query(default=None, ge=0, alias="to", description="Конец диапазона (исключая)."),
) -> Response:
    """Get текста лога, целиком или диапазоном. Любой аутентифицированный актор."""
    log, content = await svc.get_log_range(db, queue_item_id, from_, to_)
    filename = await svc.build_filename(db, log)
    return Response(
        content=content,
        media_type="text/plain; charset=utf-8",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get(
    "/{queue_item_id}/log/segments",
    response_model=PaginatedResponse[TestLogSegmentResponse],
    summary="Список чекпоинтов/команд лога",
    description=(
        "Метаданные сегментов без самого текста — офсеты для перехода к "
        "диапазону через `GET .../log?from=&to=`. `status` фильтрует "
        "(например `FATAL` — сразу видно, где упало, без чтения всего лога)."
    ),
    responses={
        401: {"description": "ACCESS_TOKEN_MISSING — запрос без bearer'а."},
        404: {"description": "TEST_LOG_NOT_FOUND — для этого queue_item лога ещё нет."},
    },
)
async def list_log_segments(
    queue_item_id: str,
    identity: AuthenticatedIdentity,
    db: AsyncSession = Depends(get_db),
    status_filter: str | None = Query(
        default=None, alias="status", description="Фильтр по статусу сегмента: OK / CHANGED / FATAL.",
    ),
    limit: int = Query(default=500, ge=1, le=2000),
    offset: int = Query(default=0, ge=0),
) -> PaginatedResponse[TestLogSegmentResponse]:
    """List сегментов лога. Любой аутентифицированный актор."""
    items, total = await svc.list_segments(db, queue_item_id, status=status_filter, limit=limit, offset=offset)
    return PaginatedResponse[TestLogSegmentResponse](
        items=[TestLogSegmentResponse.model_validate(i) for i in items],
        total=total, limit=limit, offset=offset,
    )

File: emm/web_ui/src/api/testing/testLogs.ts
/**
 * Тонкие обёртки над `testing_service` `/queue-items/{id}/log*` — чтение и
 * экспорт логов прогонов (§2.6, §8.4 плана миграции). Открыто любому
 * аутентифицированному актору, как и остальные read-пути этого сервиса.
 *
 * `WS .../log/stream` (живой лог, §8.6) — отдельный модуль `@/api/testing/logStream`.
 *
 * Source of truth: `testing_service/src/api/v1/endpoints/test_logs.py`.
 */

import { ApiError, API_BASE_URL, apiGet } from "@/api/client";
import { getAccessToken } from "@/api/tokenStore";
import type { ApiErrorPayload } from "@/api/auth/types";
import type {
  TestingPaginatedResponse,
  TestLogSegment,
} from "@/api/testing/types";

const BASE = "/testing/v1";

const CONTENT_DISPOSITION_FILENAME = /filename="?([^"]+)"?/i;

/** Диапазон текста лога — оба поля опциональны (офсеты `byte_offset_*` сегмента). */
export interface LogRange {
  from?: number;
  to?: number;
}

/** Результат `getTestLogText`/`downloadTestLog` — текст + имя файла из `Content-Disposition`. */
export interface TestLogTextResult {
  text: string;
  filename: string;
}

/**
 * `GET /queue-items/{id}/log` — весь текст лога или диапазон по офсетам
 * (`from`/`to`). Прямой `fetch` (не `apiGet`), т.к. ответ — не JSON
 * (`text/plain`) и нужен `Content-Disposition` для имени файла (§8.3:
 * `<stand>_<testname>_<os_version>_<kernel>_<date>.log`). Невалидный диапазон
 * (`from>to`, за пределами длины текста) — 422.
 */
export async function getTestLogText(
  queueItemId: string,
  range: LogRange = {},
): Promise<TestLogTextResult> {
  const params = new URLSearchParams();
  if (range.from !== undefined) params.set("from", String(range.from));
  if (range.to !== undefined) params.set("to", String(range.to));
  const qs = params.toString();
  const url = `${API_BASE_URL}${BASE}/queue-items/${queueItemId}/log${qs ? `?${qs}` : ""}`;

  const headers: Record<string, string> = { Accept: "text/plain" };
  const access = getAccessToken();
  if (access) headers["Authorization"] = `Bearer ${access}`;

  const res = await fetch(url, { method: "GET", headers, credentials: "same-origin" });

  if (!res.ok) {
    let payload: Partial<ApiErrorPayload> = {};
    try {
      payload = (await res.json()) as Partial<ApiErrorPayload>;
    } catch {
      // на ошибке эндпоинт всё равно отдаёт JSON-envelope; пустой payload —
      // fallback на случай нестандартного тела.
    }
    throw new ApiError(res.status, payload);
  }

  const disposition = res.headers.get("Content-Disposition") ?? "";
  const match = CONTENT_DISPOSITION_FILENAME.exec(disposition);
  const filename = match?.[1] ?? `${queueItemId}.log`;
  const text = await res.text();
  return { text, filename };
}

/**
 * Скачивает лог в браузере (`<a download>` + `URL.createObjectURL`) — тот же
 * приём, что и `exportEvents` в `@/api/loging/events`. Использует
 * `getTestLogText` под капотом, поэтому диапазон работает так же.
 */
export async function downloadTestLog(
  queueItemId: string,
  range: LogRange = {},
): Promise<TestLogTextResult> {
  const result = await getTestLogText(queueItemId, range);
  const blob = new Blob([result.text], { type: "text/plain;charset=utf-8" });
  const objectUrl = URL.createObjectURL(blob);
  try {
    const a = document.createElement("a");
    a.href = objectUrl;
    a.download = result.filename;
    document.body.appendChild(a);
    a.click();
    a.remove();
  } finally {
    URL.revokeObjectURL(objectUrl);
  }
  return result;
}

/** Параметры `GET /queue-items/{id}/log/segments` — фильтр по статусу + пагинация. */
export interface ListLogSegmentsQuery {
  status?: "OK" | "CHANGED" | "FATAL" | string;
  limit?: number;
  offset?: number;
}

/**
 * `GET /queue-items/{id}/log/segments` — метаданные сегментов (чекпоинты/
 * команды) без самого текста — офсеты для перехода к диапазону через
 * `getTestLogText`.
 */
export function listLogSegments(
  queueItemId: string,
  query: ListLogSegmentsQuery = {},
): Promise<TestingPaginatedResponse<TestLogSegment>> {
  return apiGet<TestingPaginatedResponse<TestLogSegment>>(
    `${BASE}/queue-items/${queueItemId}/log/segments`,
    { query: { ...query } },
  );
}
